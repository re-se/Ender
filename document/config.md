# config

```js
{
  "main": "01.end",
  "text": {
    "speed": 1000,
    "waitPeriod": false
  }
  "auto": true,
  "autoSpeed": 1000,
}
```

## main: string

一番最初に読み込む Ender スクリプト。

## text

文章関連の設定

### speed: number

文字の表示間隔。この値 ms に一文字の間隔で文字が表示される。

### waitPeriod: bool

文中の"。"で wait するかどうか。
true に設定した場合、以下のように記述することで wait をエスケープすることができる。

```
config.text.waitPeriod = true
あいう。えおか # 「う。」で停止
あいう。\えおか # 「か」まで自動で進む
```

## auto: bool

自動再生設定。true で自動再生される。

## autoSpeed: number

自動再生時の wait で止まる ms 数。

## audio

音声を鳴らす時の設定

### path: string

音声ディレクトリのパス。basePath からの相対パス指定。スクリプトから音声ファイルが呼び出されるとき、このパスからの相対パスで指定される。

### bus: BusConfig[]

音声 Bus の設定。

鳴らす音声をグルーピングすることでグループごとの音量パラメータの調整やエフェクトの一括適用などが可能になります。
このグループを bus (音声 Bus )と言います。

スクリプトから音声を鳴らす時はここで定義した音声 Bus のどれかに属した音声として再生されます。(属する音声 Bus はスクリプトから指定できます。)

また、音声にエフェクトをかけたい場合は鳴らしている音声か音声 Bus をエフェクト対象として選択することができます。

#### BusConfig

```js
{
  "name": "bgm",
  "out": "master",
  "gain": 1,
}
```

##### name: string

音声 Bus の名前です。

スクリプトから音声 Bus を指定するときや、音声 Bus の接続の関係を定義するときに使います。

#### out: string

音声 Bus の出力先設定。他の音声 Bus を指定して接続の関係を定義することができます。

空文字または未定義の場合は標準出力。

#### gain: number

音声 Bus の出力音量。0.0 <= gain <= 1.0
