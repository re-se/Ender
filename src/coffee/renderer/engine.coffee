fs = require 'fs'
path = require 'path'
{FuncEngine} = require './func'
{Config} = require './Config'
parser = require '../../../ender.js'

module.exports = class Ender
  constructor: (@Action, @config) ->
    @insts = []
    @pc = 0
    @parse(@config.main)
    @pc++
    @fe = new FuncEngine(@Action)
    @g = @_exec()
    @currentMessage = []
    @nextMessage = []
    @history = ""

  parse: (p) ->
    @filename = p
    mainPath = path.join(@config.basePath, p)
    script = fs.readFileSync(mainPath).toString()
    @insts = parser.parse(script).concat @insts[@pc+1..]
    @pc = -1

  getCurrentState: ->
    [@filename, @pc]

  setCurrentState: (filename, pc, cb) ->
    @parse filename
    @clearAll () =>
      @skip(pc, cb)

  skip: (pc, cb) ->
    _g = @g
    @g = @_skip(_g, pc, cb)
    @exec()

  _skip : (_g, pc, cb) ->
    textSpeed = @config.textSpeed
    @Action.setConfig "textSpeed", 0
    while @pc < pc
      ret = _g.next()
      if ret.value is "async"
        yield 0
    @Action.setConfig "textSpeed", textSpeed
    @g = _g
    if cb?
      cb()


  finishAnimation: ->
    @isAnimated = false

  startAnimation: ->
    @isAnimated = true

  autoExec: ->
    if @config.skip
      if @config.textSpeed > 0
        @_config = @config.dup()
        @Action.setConfig "textSpeed", 0
      else
        setTimeout =>
          @exec() if @config.skip
        , 0
      return
    else
      if @_config?
        @_config.skip = false
        @config = @_config
        @_config = null
        @Action.setConfig @config
        return
    if @config.auto
      setTimeout =>
        @exec() if @config.auto
      , @config.autoSpeed

  changeStyle: (style) ->
    @style = Object.assign {}, @style, style
    @addMessage type: "style", value: @style

  addMessage: (message) ->
    @nextMessage.push message

  showMessage: (currentMessage, nextMessage) ->
    me = @messageGen @currentMessage, @nextMessage
    if @config.textSpeed > 0
      @isTextAnimated = true
      cb = () =>
        if (m = me.next()).done
          @isTextAnimated = false
          @autoExec()
        else
          @Action.setText m.value
          @timeoutID = setTimeout cb, @config.textSpeed
      cb()
    else
      @Action.setText @addEndMarker(@currentMessage.concat @nextMessage), => @autoExec()

  clear: (type, className, effect) ->
    switch type
      when "text"
        @currentMessage = []
        @nextMessage = []
        @Action.clear("text")
      when "image"
        @Action.clearImage(className, effect)
        @isAnimated = true
      else
        @currentMessage = []
        @nextMessage = []
        @Action.clear(type)

  clearAll: (cb) ->
    @history = ""
    @clear(cb)

  _exec: =>
    @isAnimated = false
    @isTextAnimated = false
    @style = {}
    imagesMap = {}
    @pc = 0
    loop
      inst = @insts[@pc]
      next = @insts[@pc+1]
      if inst?
        console.log inst.type
        switch inst.type
          when "wait"
            @endMarker = if next?.type is "clear" then "▼" else "▽"
            @showMessage()
            yield 0
            @currentMessage = @currentMessage.concat @nextMessage
            @nextMessage = []
            if @isTextAnimated
              clearTimeout @timeoutID
              @Action.setText @addEndMarker(@currentMessage)
              @isTextAnimated = false
              yield 0
          when "text"
            for m in inst.value
              if m.type is "interpolation"
                m.type = "text"
                m.body = @fe.getVar m.expr
              @history += m.body if m.body?
              @history += "\n" if m.type is "br"
            @nextMessage = @nextMessage.concat inst.value
          when "name"
            @currentMessage = []
            @nextMessage = []
            @Action.setName(inst.name)
            @history += "#{inst.name}「"
          when "nameClear"
            @Action.setName(null)
            @history += "」\n"
          # when "br"
          #   message = message.concat type: "br"
          when "clear"
            if inst.message
              @currentMessage = []
              @nextMessage = []
              @Action.setText @currentMessage
              # @history += "\n"
            if inst.image
              @Action.clearImage(inst.target, inst.effect)
              @isAnimated = true
              yield 0 while @isAnimated
          when "funcdecl"
            @fe.addFunc(inst)
          when "func"
            it = @fe.exec(@)
            while not (ret = it.next()).done
              yield ret.value
          else
            console.error inst
      @pc++
      break if @insts.length < @pc

    @clearAll()
    @g = @_exec()

  exec: =>
    @g.next()

  addEndMarker: (message, isNewPage) ->
    endMarker = {type: "marker"}
    endMarker.body = @endMarker
    last = message[message.length - 1]
    if last.type is "br"
      message[0..message.length - 2].concat endMarker, last
    else
      message.concat endMarker

  messageGen: (before, message) ->
    i = 0
    for m in message
      cur = {}
      cur.type = m.type
      j = 1
      if m.body?
        while j <= m.body.length
          cur.body = m.body.substr(0, j)
          yield before.concat message.slice(0, i), cur
          j++
      i++
    yield @addEndMarker before.concat(message)
