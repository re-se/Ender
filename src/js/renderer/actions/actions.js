// root
export const resetState = (components) => {
  return { type: 'RESET_STATE' }
}

// components
export const addComponents = (components) => {
  return {
    type: 'ADD_COMPONENTS',
    components
  }
}

// configPath
export const setConfigPath = (path) => {
  return {
    type: 'SET_CONFIG_PATH',
    path
  }
}

// MessageBox
export const setMessage = (message) => {
  return {
    type: 'SET_MESSAGE',
    message
  }
}

export const setMessagePosition = ({index, position}) => {
  return {
    type: 'SET_MESSAGE_POSITION',
    index,
    position
  }
}

export const setMessageClassNames = (classNames) => {
  return {
    type: 'SET_MESSAGE_CLASSNAMES',
    classNames
  }
}

// animation
export const startAnimation = (animation) => {
  return {
    type: 'START_ANIMATION',
    animation
  }
}

export const finishAnimation = () => {
  return {
    type: 'FINISH_ANIMATION',
  }
}

export const addImage = (image) => {
  return {
    type: 'ADD_IMAGE',
    image
  }
}
