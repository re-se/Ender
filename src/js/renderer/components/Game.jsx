import React from 'react';
import generateComponent from '../utils/generateComponent';

export type Props = {
  children: ComponentMaterial
};

const Game = ({children}: Props) => {
  <div id="inner" className="inner-view" key="inner-view">
   {generateComponent(children)}
  </div>
}

export default Game;
