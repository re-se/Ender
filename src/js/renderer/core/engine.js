//@flow

import fs from 'fs';
import path from 'path';
import parser from './ender.js';
import instMap from './instMap.js';

class Ender {
  insts: Inst[];
  pc: number;

  constructor(scriptPath: string) {
    this._loadScript(scriptPath);
    this.pc = 0;
  }

  /**
   * スクリプトファイルを読み込む
   * @param {string} scriptPath スクリプトファイルのフルパス
   */
  _loadScript(scriptPath: string) {
    let script = fs.readFileSync(scriptPath).toString();
    this.insts = parser.parse(script);
  }

  /**
   * 命令列を実行する
   * @return {[type]} [description]
   */
  exec() {
    let inst = this.insts[this.pc];
    // 命令実行
    instMap[inst.type](this, inst);

    this.pc++;
  }
}

export default Ender;
