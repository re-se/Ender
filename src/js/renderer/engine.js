// @flow
import fs from 'fs';
import path from 'path';
import FuncEngine from './func';
import { Config } from './Config';
import parser from './ender.js';

export default class Ender {
  pc: number;
  insts: Array<Inst>;
  config: Config;
  endMarker: string;
  constructor(Action, config) {
    this._exec = this._exec.bind(this);
    this.exec = this.exec.bind(this);
    this.Action = Action;
    this.config = config;
    this.insts = [];
    this.pc = 0;
    this.fe = new FuncEngine(this.Action);
    this.load(this.config.main);
  }

  load(filename, cb) {
    this.filename = filename;
    this.parse();
    this.pc = 0;
    this.g = this._exec();
    this.clearAll(cb);
  }

  reload : (filename: string, cb: Function) => void;
  reload : (filename: Function) => void;
  reload(filename, cb) {
    if (typeof filename !== "string") {
      cb = filename;
      ({ filename } = this);
    }
    this.insts = [];
    this.pc = 0;
    this.load(filename, cb);
  }

  parse() {
    let textPath = this.config.text.path;
    if (textPath == null) { textPath = ""; }
    const mainPath = path.join(this.config.basePath, textPath, this.filename);
    const script = fs.readFileSync(mainPath).toString();
    return this.insts = parser.parse(script).concat(this.insts.slice(this.pc+1));
  }

  getCurrentState() {
    return [this.filename, this.pc];
  }

  setCurrentState(filename, pc, cb) {
    return this.reload(filename, () => {
      return this.skip(pc, cb);
    });
  }

  skip(pc, cb) {
    if (this.isSkip) { return; }
    this.isSkip = true;
    let _g = this.g;
    let callback = () => {
      this.isSkip = false;
      return cb();
    };
    this.g = this._skip(_g, pc, callback);
    return this.exec();
  }

  *_skip(_g, pc, cb) {
    let { textSpeed } = this.config;
    this.Action.setConfig("textSpeed", 0);
    let audioVolume = this.config.volume.Master;
    let muteVolume = this.config.volume;
    muteVolume.Master = 0;
    this.Action.setConfig("volume", muteVolume);
    console.log(this.config);
    while ((this.pc < pc) || this.config.skip) {
      let ret = _g.next();
      console.log(ret);
      if (ret.value === "async") {
        yield 0;
      }
    }
    muteVolume.Master = audioVolume;
    this.Action.setConfig("volume", muteVolume);
    this.Action.setConfig("textSpeed", textSpeed);
    this.g = _g;
    if (cb != null) {
      return cb();
    }
  }

  finishAnimation() {
    this.isAnimated = false;
  }

  startAnimation() {
    this.isAnimated = true;
  }

  finishLoad() {
    this.isLoaded = false;
  }

  startLoad() {
    this.isLoaded = true;
  }

  changeStyle(style) {
    this.style = Object.assign({}, this.style, style);
    this.addMessage({type: "style", value: this.style});
  }

  addMessage(message) {
    this.nextMessage.push(message);
  }

  showMessage(currentMessage, nextMessage) {
    let me = this.messageGen(this.currentMessage, this.nextMessage);
    if (this.config.textSpeed > 0) {
      this.isTextAnimated = true;
      var cb = () => {
        let m;
        if ((m = me.next()).done) {
          this.isTextAnimated = false;
          return this.autoExec();
        } else {
          this.Action.setText(m.value);
          return this.timeoutID = setTimeout(cb, this.config.textSpeed);
        }
      };
      return cb();
    } else {
      return this.Action.setText(this.addEndMarker(this.currentMessage.concat(this.nextMessage)), () => this.autoExec());
    }
  }

  clear(type, className, effect, cb) {
    switch (type) {
      case "text":
        this.currentMessage = [];
        this.nextMessage = [];
        return this.Action.clear("text", cb);
      case "image":
        return this.Action.clearImage(className, effect, cb);
      default:
        this.currentMessage = [];
        this.nextMessage = [];
        this.history = "";
        cb = type;
        return this.Action.clear(cb);
    }
  }

  clearAll(cb) {
    this.history = "";
    return this.clear(cb);
  }

  *_exec() {
    this.isAnimated = false;
    this.isTextAnimated = false;
    this.isLoaded = false;
    this.style = {};
    let imagesMap = {};
    this.pc = 0;
    while (true) {
      let inst = this.insts[this.pc];
      let next = this.insts[this.pc+1];
      if (inst != null) {
        var ret;
        console.log(inst.type);
        switch (inst.type) {
          case "wait":
            this.endMarker = (next != null ? next.type : undefined) === "clear" ? "▼" : "▽";
            this.showMessage();
            yield 0;
            this.currentMessage = this.currentMessage.concat(this.nextMessage);
            this.nextMessage = [];
            if (this.isTextAnimated) {
              clearTimeout(this.timeoutID);
              this.Action.setText(this.addEndMarker(this.currentMessage));
              this.isTextAnimated = false;
              yield 0;
            }
            break;
          case "text":
            for (let m of Array.from(inst.value)) {
              if (m.type === "interpolation") {
                m.type = "text";
                m.body = this.fe.getVar(m.expr);
              }
              if (m.body != null) { this.history += m.body; }
              if (m.type === "br") { this.history += "\n"; }
            }
            this.nextMessage = this.nextMessage.concat(inst.value);
            break;
          case "name":
            this.currentMessage = [];
            this.nextMessage = [];
            let { name } = inst;
            if ((name.length < 1) && (this.config.text.defaultName != null)) {
              name = this.config.text.defaultName;
            }
            this.Action.setName(name);
            this.history += `${name}「`;
            break;
          case "nameClear":
            this.Action.setName(null);
            this.history += "」\n";
            break;
          // when "br"
          //   message = message.concat type: "br"
          case "clear":
            if (inst.message) {
              this.clear("text", () => {
                return this.Action.setText(this.currentMessage);
              }
              );
            } else {
              console.error(inst);
            }
            break;
              // @history += "\n"
            // if inst.image
            //   console.error "wawawa"
            //   console.dir inst
            //   @Action.clearImage(inst.target, inst.effect, @exec)
            //   @startAnimation()
            //   loop
            //     yield "async"
            //     break if not @isAnimated
          case "funcdecl":
            this.fe.addFunc(inst);
            break;
          case "func":
            let it = this.fe.exec(this);
            while (!(ret = it.next()).done) {
              yield ret.value;
            }
            break;
          default:
            console.error(inst);
        }
      }
      this.pc++;
      if (this.insts.length < this.pc) { break; }
    }

    this.clearAll();
    return this.g = this._exec();
  }

  exec() {
    try {
      let ret = this.g.next();
      return this.isAsync = ret.value === "async";
    } catch (error) {
      console.error(error);
    }
  }

  autoExec() {
    if (this.config.auto && !this.isAsync) {
      setTimeout(() => {
        if (this.config.auto && !this.isAsync) {
          return this.exec();
        }
      }, this.config.autoSpeed);
    }
  }


  addEndMarker(message, isNewPage) {
    let endMarker = {type: "marker"};
    endMarker.body = this.endMarker;
    let last = message[message.length - 1];
    if (last.type === "br") {
      return message.slice(0, message.length - 2 + 1 || undefined).concat(endMarker, last);
    } else {
      return message.concat(endMarker);
    }
  }

  *messageGen(before, message) {
    let i = 0;
    for (let m of Array.from(message)) {
      let cur = {};
      cur.type = m.type;
      let j = 1;
      if (m.body != null) {
        while (j <= m.body.length) {
          cur.body = m.body.substr(0, j);
          yield before.concat(message.slice(0, i), cur);
          j++;
        }
      }
      i++;
    }
    yield this.addEndMarker(before.concat(message));
  }
};

module.exports = Ender;
