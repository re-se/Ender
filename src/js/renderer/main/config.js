import path from 'path'
import Config from '../util/Config'
import store from './store'
import { toAbsolutePath, isExistFile, isDevelop } from '../util/util'
import fs from 'fs'

export function generateConfig(configPath: string): Config {
  let config = Config.generateConfig(configPath)
  config.auto = false
  config.extendGetter('basePath', function(key, getter) {
    return function() {
      let basePath = getter() || path.dirname(this._path)
      return toAbsolutePath(basePath)
    }
  })
  return config
}

export function saveConfig(config: Config) {
  // 書き込み予定のセーブファイルがすでにあったら削除
  const configFilePath = path.join(config.basePath, 'config.json')
  if (isExistFile(configFilePath)) {
    fs.unlink(configFilePath, err => {
      if (err) {
        console.warn(err)
      }
    })
  }

  fs.writeFile(configFilePath, config.toString(), err => {
    if (err) {
      console.warn(err)
    } else if (isDevelop()) {
      console.log(`done save config. ${configFilePath}`)
    }
  })
}
