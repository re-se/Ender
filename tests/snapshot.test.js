// @flow
import { ipcRenderer } from 'electron'
import React from 'react'
import { Provider } from 'react-redux'

import Game from '../src/js/renderer/components/Game'
import engine from '../src/js/renderer/main/engine'
import { toAbsolutePath } from '../src/js/renderer/util/util'
import store from '../src/js/renderer/main/store'
import { generateConfig } from '../src/js/renderer/main/config'
import { resetState } from '../src/js/renderer/actions/actions'

import fs from 'fs'
import path from 'path'

import renderer from 'react-test-renderer'

const testDirnames = fs.readdirSync(__dirname)
testDirnames.forEach(testName => {
  if (testName.startsWith('__')) {
    return
  }
  if (!fs.statSync(path.join(__dirname, testName)).isDirectory()) {
    return
  }
  describe('components test', () => {
    beforeAll(() => {
      // ruby で width を求めるあたりで使っているものの mock
      const createElement = document.createElement.bind(document)
      document.createElement = tagName => {
        if (tagName === 'canvas') {
          return {
            getContext: () => ({
              measureText: () => ({}),
            }),
          }
        }
        return createElement(tagName)
      }
    })
    it(`${testName}`, () => {
      let configPath = path.join(__dirname, testName, 'config.json')
      const config = generateConfig(configPath)
      store.dispatch(resetState())
      engine.init(config)

      const tree = renderer.create(
        <Provider store={store}>
          <Game />
        </Provider>,
        document.getElementById('contents')
      )
      while (!engine.isFinished) {
        debugger
        const inst = engine.lookahead()
        if (engine._yieldValue === 'snapshot') {
          debugger
          expect(tree.toJSON()).toMatchSnapshot()
        }
        engine.exec()
      }
    })
  })
})
