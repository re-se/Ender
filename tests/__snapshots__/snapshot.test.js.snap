// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`components test import 1`] = `
<div
  className="inner-view"
  id="inner"
>
  <div
    className="ender-box "
  >
    <style>
      display: block;
    </style>
  </div>
  <div
    className="ender-box "
  >
    <div
      className="ender-box test1"
    />
  </div>
</div>
`;

exports[`components test layout_base 1`] = `
<div
  className="inner-view"
  id="inner"
>
  <div
    className="ender-box "
  >
    <div
      className="ender-box sample"
    />
  </div>
</div>
`;

exports[`components test literal 1`] = `
<div
  className="inner-view"
  id="inner"
>
  <div
    className="ender-box "
  >
    <div
      className="ender-box "
    >
      <div
        className="ender-messageBox "
      >
        <div
          className="ender-messageBox-inner"
        >
          <span
            style={Object {}}
          >
            a: 
          </span>
          <span
            style={Object {}}
          >
            1
          </span>
          <br
            style={Object {}}
          />
          <span
            style={Object {}}
          >
            b: {a: 
          </span>
          <span
            style={Object {}}
          >
            1
          </span>
          <span
            style={Object {}}
          >
            }
          </span>
          <br
            style={Object {}}
          />
          <span
            style={Object {}}
          >
            c: [
          </span>
          <span
            style={Object {}}
          >
            1
          </span>
          <span
            style={Object {}}
          >
            , 
          </span>
          <span
            style={Object {}}
          >
            1
          </span>
          <span
            style={Object {}}
          >
            ]
          </span>
          <br
            style={Object {}}
          />
          <span
            className="marker"
          >
            ▼
          </span>
        </div>
      </div>
    </div>
    <div
      className="full"
      onClick={[Function]}
    />
  </div>
</div>
`;

exports[`components test message_box 1`] = `
<div
  className="inner-view"
  id="inner"
>
  <div
    className="ender-box "
  >
    <div
      className="ender-box "
    >
      <div
        className="ender-messageBox "
      >
        <div
          className="ender-messageBox-inner"
        >
          <span
            style={Object {}}
          >
            aaa
          </span>
          <br
            style={Object {}}
          />
          <span
            style={Object {}}
          >
            bbb
          </span>
          <br
            style={Object {}}
          />
          <span
            className="ender-ruby"
          >
            <span
              className="ender-rt"
              style={Object {}}
            >
              <span
                style={
                  Object {
                    "opacity": 1,
                  }
                }
              >
                ﹅
              </span>
              <span
                style={
                  Object {
                    "opacity": 1,
                  }
                }
              >
                ﹅
              </span>
              <span
                style={
                  Object {
                    "opacity": 1,
                  }
                }
              >
                ﹅
              </span>
            </span>
            <span
              className="ender-rb"
              style={Object {}}
            >
              <span
                style={
                  Object {
                    "opacity": 1,
                  }
                }
              >
                c
              </span>
              <span
                style={
                  Object {
                    "opacity": 1,
                  }
                }
              >
                c
              </span>
              <span
                style={
                  Object {
                    "opacity": 1,
                  }
                }
              >
                c
              </span>
            </span>
          </span>
          <br
            style={Object {}}
          />
          <span
            className="ender-ruby"
          >
            <span
              className="ender-rt"
              style={Object {}}
            >
              <span
                style={
                  Object {
                    "opacity": 1,
                  }
                }
              >
                e
              </span>
              <span
                style={
                  Object {
                    "opacity": 1,
                  }
                }
              >
                e
              </span>
              <span
                style={
                  Object {
                    "opacity": 1,
                  }
                }
              >
                e
              </span>
            </span>
            <span
              className="ender-rb"
              style={Object {}}
            >
              <span
                style={
                  Object {
                    "opacity": 1,
                  }
                }
              >
                d
              </span>
              <span
                style={
                  Object {
                    "opacity": 1,
                  }
                }
              >
                d
              </span>
              <span
                style={
                  Object {
                    "opacity": 1,
                  }
                }
              >
                d
              </span>
            </span>
          </span>
          <br
            style={Object {}}
          />
          <span
            style={Object {}}
          >
            fff
          </span>
          <span
            style={Object {}}
          >
            ggg
          </span>
          <br
            style={Object {}}
          />
          <span
            style={Object {}}
          >
            hhh
          </span>
          <br
            style={Object {}}
          />
          <span
            style={Object {}}
          >
            iii
          </span>
          <br
            style={Object {}}
          />
          <span
            className="marker"
          >
            ▼
          </span>
        </div>
      </div>
    </div>
    <div
      className="full"
      onClick={[Function]}
    />
  </div>
</div>
`;

exports[`components test message_box 2`] = `
<div
  className="inner-view"
  id="inner"
>
  <div
    className="ender-box "
  >
    <div
      className="ender-box "
    >
      <div
        className="ender-messageBox "
      >
        <div
          className="ender-messageBox-inner"
        >
          <span
            style={Object {}}
          >
            あああ
          </span>
          <span
            style={Object {}}
          >
            。
          </span>
          <span
            style={Object {}}
          >
            
          </span>
          <span
            style={Object {}}
          >
            いいい
          </span>
          <span
            style={Object {}}
          >
            ううう
          </span>
          <br
            style={Object {}}
          />
          <span
            style={Object {}}
          >
            えええ。おおお
          </span>
          <br
            style={Object {}}
          />
          <span
            className="marker"
          >
            ▼
          </span>
        </div>
      </div>
    </div>
    <div
      className="full"
      onClick={[Function]}
    />
  </div>
</div>
`;

exports[`components test message_box_ruby 1`] = `
<div
  className="inner-view"
  id="inner"
>
  <div
    className="ender-box "
  >
    <div
      className="ender-box "
    >
      <div
        className="ender-messageBox "
      >
        <div
          className="ender-messageBox-inner"
        >
          <span
            style={Object {}}
          >
            前
          </span>
          <span
            className="ender-ruby"
          >
            <span
              className="ender-rt"
              style={Object {}}
            >
              <span
                style={
                  Object {
                    "opacity": 1,
                  }
                }
              >
                ﹅
              </span>
              <span
                style={
                  Object {
                    "opacity": 1,
                  }
                }
              >
                ﹅
              </span>
              <span
                style={
                  Object {
                    "opacity": 1,
                  }
                }
              >
                ﹅
              </span>
            </span>
            <span
              className="ender-rb"
              style={Object {}}
            >
              <span
                style={
                  Object {
                    "opacity": 1,
                  }
                }
              >
                あ
              </span>
              <span
                style={
                  Object {
                    "opacity": 1,
                  }
                }
              >
                あ
              </span>
              <span
                style={
                  Object {
                    "opacity": 1,
                  }
                }
              >
                あ
              </span>
            </span>
          </span>
          <span
            style={Object {}}
          >
            後
          </span>
          <br
            style={Object {}}
          />
          <span
            style={Object {}}
          >
            前
          </span>
          <span
            className="ender-ruby"
          >
            <span
              className="ender-rt"
              style={Object {}}
            >
              <span
                style={
                  Object {
                    "opacity": 1,
                  }
                }
              >
                い
              </span>
              <span
                style={
                  Object {
                    "opacity": 1,
                  }
                }
              >
                い
              </span>
              <span
                style={
                  Object {
                    "opacity": 1,
                  }
                }
              >
                い
              </span>
            </span>
            <span
              className="ender-rb"
              style={Object {}}
            >
              <span
                style={
                  Object {
                    "opacity": 1,
                  }
                }
              >
                あ
              </span>
              <span
                style={
                  Object {
                    "opacity": 1,
                  }
                }
              >
                あ
              </span>
              <span
                style={
                  Object {
                    "opacity": 1,
                  }
                }
              >
                あ
              </span>
            </span>
          </span>
          <span
            style={Object {}}
          >
            後
          </span>
          <br
            style={Object {}}
          />
          <span
            style={Object {}}
          >
            前
          </span>
          <span
            className="ender-ruby"
          >
            <span
              className="ender-rt"
              style={Object {}}
            >
              <span
                style={
                  Object {
                    "opacity": 1,
                  }
                }
              >
                い
              </span>
              <span
                style={
                  Object {
                    "opacity": 1,
                  }
                }
              >
                い
              </span>
              <span
                style={
                  Object {
                    "opacity": 1,
                  }
                }
              >
                い
              </span>
            </span>
            <span
              className="ender-rb"
              style={Object {}}
            >
              <span
                style={
                  Object {
                    "opacity": 1,
                  }
                }
              >
                あ
              </span>
              <span
                style={
                  Object {
                    "opacity": 1,
                  }
                }
              >
                あ
              </span>
              <span
                style={
                  Object {
                    "opacity": 1,
                  }
                }
              >
                あ
              </span>
              <span
                style={
                  Object {
                    "opacity": 1,
                  }
                }
              >
                あ
              </span>
              <span
                style={
                  Object {
                    "opacity": 1,
                  }
                }
              >
                あ
              </span>
              <span
                style={
                  Object {
                    "opacity": 1,
                  }
                }
              >
                あ
              </span>
              <span
                style={
                  Object {
                    "opacity": 1,
                  }
                }
              >
                あ
              </span>
              <span
                style={
                  Object {
                    "opacity": 1,
                  }
                }
              >
                あ
              </span>
              <span
                style={
                  Object {
                    "opacity": 1,
                  }
                }
              >
                あ
              </span>
            </span>
          </span>
          <span
            style={Object {}}
          >
            後
          </span>
          <br
            style={Object {}}
          />
          <span
            style={Object {}}
          >
            前
          </span>
          <span
            className="ender-ruby"
          >
            <span
              className="ender-rt"
              style={Object {}}
            >
              <span
                style={
                  Object {
                    "opacity": 1,
                  }
                }
              >
                い
              </span>
              <span
                style={
                  Object {
                    "opacity": 1,
                  }
                }
              >
                い
              </span>
              <span
                style={
                  Object {
                    "opacity": 1,
                  }
                }
              >
                い
              </span>
              <span
                style={
                  Object {
                    "opacity": 1,
                  }
                }
              >
                い
              </span>
              <span
                style={
                  Object {
                    "opacity": 1,
                  }
                }
              >
                い
              </span>
              <span
                style={
                  Object {
                    "opacity": 1,
                  }
                }
              >
                い
              </span>
              <span
                style={
                  Object {
                    "opacity": 1,
                  }
                }
              >
                い
              </span>
              <span
                style={
                  Object {
                    "opacity": 1,
                  }
                }
              >
                い
              </span>
              <span
                style={
                  Object {
                    "opacity": 1,
                  }
                }
              >
                い
              </span>
            </span>
            <span
              className="ender-rb"
              style={Object {}}
            >
              <span
                style={
                  Object {
                    "opacity": 1,
                  }
                }
              >
                あ
              </span>
              <span
                style={
                  Object {
                    "opacity": 1,
                  }
                }
              >
                あ
              </span>
              <span
                style={
                  Object {
                    "opacity": 1,
                  }
                }
              >
                あ
              </span>
            </span>
          </span>
          <span
            style={Object {}}
          >
            後
          </span>
          <br
            style={Object {}}
          />
          <span
            className="marker"
          >
            ▼
          </span>
        </div>
      </div>
    </div>
    <div
      className="full"
      onClick={[Function]}
    />
  </div>
</div>
`;

exports[`components test set_return_value 1`] = `
<div
  className="inner-view"
  id="inner"
>
  <div
    className="ender-box "
  >
    <div
      className="ender-box "
    >
      <div
        className="ender-messageBox "
      >
        <div
          className="ender-messageBox-inner"
        >
          <span
            style={Object {}}
          >
            1
          </span>
          <br
            style={Object {}}
          />
          <span
            className="marker"
          >
            ▼
          </span>
        </div>
      </div>
    </div>
    <div
      className="full"
      onClick={[Function]}
    />
  </div>
</div>
`;

exports[`components test variable 1`] = `
<div
  className="inner-view"
  id="inner"
>
  <div
    className="ender-box "
  >
    <div
      className="ender-messageBox "
    >
      <div
        className="ender-messageBox-inner"
      >
        <span
          style={Object {}}
        >
          test
        </span>
        <br
          style={Object {}}
        />
        <span
          style={Object {}}
        >
          string
        </span>
        <br
          style={Object {}}
        />
        <span
          className="marker"
        >
          ▼
        </span>
      </div>
    </div>
    <div
      className="full"
      onClick={[Function]}
    />
  </div>
</div>
`;
